pkg/sql/catalog/descs/hydrate.go

```diff
@@ -12,369 +12,165 @@ package descs
+// leasing depending on the context. This is set via the flags.
+//
+// Collection method callers expect the descriptors to come back hydrated.
+// In practice, array types here are not hydrated, and that's a bummer.
+// Nobody presently is upset about it, but it's not a good thing.
+// Ideally we'd have a clearer contract regarding hydration and the values
+// stored in the various maps inside the collection. One might want to
+// store only hydrated values in the various maps. This turns out to be
+// somewhat tricky because we'd need to make sure to properly re-hydrate
+// all the relevant descriptors when a type descriptor change. Leased
+// descriptors are at least as tricky, plus, there we have a cache that
+// works relatively well.
+//
+// TODO(ajwerner): Sort out the hydration mess; define clearly what is
+// hydrated where and test the API boundary accordingly.
+func (tc *Collection) hydrateDescriptors(
+	ctx context.Context, txn *kv.Txn, flags tree.CommonLookupFlags, descs []catalog.Descriptor,
+) error {
+
+	var hydratableMutableIndexes, hydratableImmutableIndexes []int
+	for i, desc := range descs {
+		if desc == nil || !hydrateddesc.IsHydratable(desc) {
+			continue
+		}
+		if _, ok := desc.(catalog.MutableDescriptor); ok {
+			hydratableMutableIndexes = append(hydratableMutableIndexes, i)
+		} else {
+			hydratableImmutableIndexes = append(hydratableImmutableIndexes, i)
 		}
-		// The cache decided not to give back a hydrated descriptor, likely
-		// because either we've modified the table or one of the types or because
-		// this transaction has a stale view of one of the relevant descriptors.
-		// Proceed to hydrating a fresh copy.
 	}
 
-	// ImmutableTableDescriptors need to be copied before hydration, because
-	// they are potentially read by multiple threads.
-	getType := getImmutableTypeLookupFunc(tc, txn, includeOffline, avoidLeased, sc)
-	switch t := desc.(type) {
-	case catalog.TableDescriptor:
-		mut := t.NewBuilder().BuildExistingMutable().(*tabledesc.Mutable)
-		if err := typedesc.HydrateTypesInTableDescriptor(ctx, mut.TableDesc(), getType); err != nil {
-			return nil, err
-		}
-		return mut.ImmutableCopy().(catalog.TableDescriptor), nil
-	case catalog.SchemaDescriptor:
-		mut := t.NewBuilder().BuildExistingMutable().(*schemadesc.Mutable)
-		if err := typedesc.HydrateTypesInSchemaDescriptor(ctx, mut.SchemaDesc(), getType); err != nil {
-			return nil, err
-		}
-		return mut.ImmutableCopy().(catalog.SchemaDescriptor), nil
-	case catalog.FunctionDescriptor:
-		mut := t.NewBuilder().BuildExistingMutable().(*funcdesc.Mutable)
-		if err := typedesc.HydrateTypesInFunctionDescriptor(ctx, mut.FuncDesc(), getType); err != nil {
-			return nil, err
+	// Hydrate mutable hydratable descriptors of the slice in-place.
+	if len(hydratableMutableIndexes) > 0 {
+		typeFn := makeMutableTypeLookupFunc(tc, txn, descs)
+		for _, i := range hydratableMutableIndexes {
+			if err := hydrateddesc.Hydrate(ctx, descs[i], typeFn); err != nil {
```

Comment:
I've hoisted a lot of the hydration logic to `hydrateddesc` to unify both the `getDescriptorsByID` and `nstree.Catalog` use cases.